File: cs50x/week6/pset/sentimental-credit/credit.py
# TODO


def main():
    number = get_pos()
    unmultiplied = number
    multiplied = number

    # Check Card Lengh
    ccNUM3 = number
    count = 0
    while ccNUM3 >= 1:
        ccNUM3 = ccNUM3 / 10
        count += 1

    # Luna's Algorithm
    # second digits
    second = multiplied / 10
    total_multi = 0
    first_sec = int(second % 10) * 2

    while second >= 1:
        second = second / 100
        multi = int(second % 10) * 2
        if multi > 9:
            multi_digit_a = int(multi % 10)
            multi_digit_b = int(multi / 10)
            total_multi = total_multi + multi_digit_a + multi_digit_b
        else:
            total_multi = total_multi + multi
    if first_sec > 9:
        first_digit_a = int(first_sec % 10)
        first_digit_b = int(first_sec / 10)
        total_multi2 = total_multi + first_digit_a + first_digit_b
    else:
        total_multi2 = first_sec + total_multi

    # first digits
    first_digit = int(unmultiplied % 10)
    total_sum = 0

    while unmultiplied >= 1:
        unmultiplied = unmultiplied / 100
        sum = int(unmultiplied % 10)
        total_sum = total_sum + sum
    total_sum = total_sum + first_digit

    all = total_multi2 + total_sum

    if all % 10 == 0:
        if (
            count == 13
            or count == 16
            and int(number / (10**12)) == 4
            or int(number / (10**15)) == 4
        ):
            print("VISA")
        elif (
            count == 16
            and int(number / (10**14)) == 51
            or int(number / (10**14)) == 52
            or int(number / (10**14)) == 53
            or int(number / (10**14)) == 54
            or int(number / (10**14)) == 55
        ):
            print("MASTERCARD")
        elif (
            count == 15
            and int(number / (10**13)) == 34
            or int(number / (10**13)) == 37
        ):
            print("AMEX")
        else:
            print("INVALID")
    else:
        print("INVALID")


def get_pos():
    while True:
        try:
            n = int(input("Number: "))
            if n > 0:
                return n
        except ValueError:
            print("Not an intger.")


main()

File: cs50x/week1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

long getpos(void);
int main(void)
{
    // Promp the card number from the user
    long number = getpos();
    long unmultiplied = number;
    long multiplied = number;
    long ccNUM3 = number;

    // Check Card Lengh
    int count = 0;
    for (int i = 0; ccNUM3 > 0; i++)
    {
        ccNUM3 = ccNUM3 / 10;
        count++;
    }
    // printf("count: %i\n", count);

    // Luhn’s Algorithm
    //  Multiply every other digit by 2, starting with the number’s second-to-last digit, and then add those products’ digits
    //  together.
    int multitotal = 0;
    long digit2;
    long multiplied2 = multiplied / 10;
    while (multiplied2 > 0)
    {
        multiplied2 = multiplied2 / 100;
        digit2 = multiplied2 % 10;
        int multidigit2 = digit2 * 2;
        int a;
        int b;
        if (multidigit2 > 9)
        {
            a = multidigit2 % 10;
            b = (multidigit2 / 10) % 10;
            multitotal = multitotal + a + b;
        }
        else
        {
            multitotal = multitotal + multidigit2;
        }
    }
    long lastdigit2 = (multiplied / 10) % 10;
    long multilastdigit2 = lastdigit2 * 2;
    int c;
    int d;
    int multitotal2;
    if (multilastdigit2 > 9)
    {
        c = multilastdigit2 % 10;
        d = (multilastdigit2 / 10) % 10;
        multitotal2 = multitotal + c + d;
    }
    else
    {
        multitotal2 = multitotal + multilastdigit2;
    }

    // printf("%i\n", multitotal);

    // Collect the unmultiplied numbers
    long lastdigit = number % 10;
    int multitotal3 = 0;
    long digit1;
    while (unmultiplied > 0)
    {
        unmultiplied = unmultiplied / 100;
        digit1 = unmultiplied % 10;
        multitotal3 = multitotal3 + digit1;
    }
    int multitotal4 = multitotal3 + lastdigit;

    // Add the sum to the sum of the digits that weren’t multiplied by 2
    int sum = multitotal2 + multitotal4;
    // printf("sum: %i\n", sum);

    // Check the type of the card
    if (sum % 10 == 0)
    {
        if (count == 16 && (number / 100000000000000 == 51 || number / 100000000000000 == 52 || number / 100000000000000 == 52 ||
                            number / 100000000000000 == 53 || number / 100000000000000 == 54 || number / 100000000000000 == 55))
        {
            printf("MASTERCARD\n");
        }

        else if ((count == 13 && number / 1000000000000 == 4) || (count == 16 && number / 1000000000000000 == 4))
        {
            printf("VISA\n");
        }

        else if (count == 15 && (number / 10000000000000 == 34 || number / 10000000000000 == 37))
        {
            printf("AMEX\n");
        }

        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

long x;
long getpos(void)
{
    do
    {
        x = get_long("Number: ");
    }
    while (x < 0);
    return x;
}